### ISSUE TITLE
many and takeWhile crash in combination

### ISSUE THREAD

**Issue by l2i** [tpolecat/atto#82]
There is no termination, once the compiler starts on
`val la = many(takeWhile(c => c != 'r')).parseOnly("awe")`.

**Comment by tpolecat** [comment]
Yeah `many` doesn't work with parsers that can successfully consume zero characters; you can parse an infinite number of empty strings. I'm not sure how to fix this without distinguishing between empty and non-empty parsers at the type level.
